Move RVFI to C

trdthg opened the pull request:
- Move the struct definitions to C structs
- Move sail rvfi registers to C global variables
- Move callback functions to C functions
- Create sail api to access C global variables

---

resolve part of #643 
completely untested, waiting for #933.

- _rvfi_dii.sail_ has been completely rewritten. it's only used to declare some C functions that sail needs to access
- _rvfi_dii.h_ now contains all the struct and function definitions
- The current layout of _rvfi_dii.c_:
	1. Initialize rvfi related global variables
	2. Functions for creating empty structure. If you use the syntax `FOO foo = { 0 }`, there will be warnings
	3. Functions needed by `rvfi_handler` like `get_packet_v1/v2` ...
	5. Functions for callbacks
	6. Functions that need to be called by sail (these functions' headers are defined in _rvfi_dii_sail.h_ seperately to avoid C importing C++ class)
	7. Implementation of the original `Class rvfi_handler`

~The code is still a bit rough now, and the correctness still needs to wait for testing.~ TestRIG passed now

Currently, any suggestions for simplify, naming, file splitting, etc. are welcome

arichardson commented on c_emulator/rvfi_dii.cpp:
```diff
@@ -16,6 +16,376 @@
+RVFI_DII_Instruction_Packet Mk_RVFI_DII_Instruction_Packet()
+{
+  return RVFI_DII_Instruction_Packet {
+      .rvfi_insn = 0,
```
Since it's C++ we can now add default initializers to the struct so there is no need for this.

arichardson commented on c_emulator/rvfi_dii.h:
```diff
@@ -23,7 +23,241 @@ class rvfi_handler {
+} RVFI_DII_Execution_Packet_V1;
+
+typedef struct /* bits(512) */ {
+  uint64_t magic;      // 63 .. 0, must be set to 'trace-v2'
```
```suggestion
  char magic[8] = {'t', 'r', 'a', 'c', 'e', '-', 'v', '2'};      // 63 .. 0, must be set to 'trace-v2'
```

arichardson commented on c_emulator/rvfi_dii.h:
```diff
@@ -23,7 +23,241 @@ class rvfi_handler {
+} RVFI_DII_Execution_Packet_PC;
+
+typedef struct {
+  uint64_t magic; // 63 .. 0, must be "int-data"
+  /// rvfi_rd_wdata is the value of the x register addressed by rd after
+  /// execution of this instruction. This output must be zero when rd is zero.
+  uint64_t rvfi_rd_wdata; // 127 ..  64
+  /// rvfi_rs1_rdata/rvfi_rs2_rdata is the value of the x register addressed
+  /// by rs1/rs2 before execution of this instruction. This output must be
+  /// zero when rs1/rs2 is zero.
+  uint64_t rvfi_rs1_rdata; // 191 .. 128
+  uint64_t rvfi_rs2_rdata; // 255 .. 192
+  /// rvfi_rd_addr is the decoded rd register address for the retired
+  /// instruction. For an instruction that writes no rd register, this output
+  /// must always be zero.
+  uint8_t rvfi_rd_addr; // 263 .. 256
+  /// rvfi_rs1_addr and rvfi_rs2_addr are the decoded rs1 and rs1 register
+  /// addresses for the retired instruction. For an instruction that reads no
+  /// rs1/rs2 register, this output can have an arbitrary value. However, if
+  /// this output is nonzero then rvfi_rs1_rdata must carry the value stored
+  /// in that register in the pre-state.
+  uint8_t rvfi_rs1_addr; // 271 .. 264
+  uint8_t rvfi_rs2_addr; // 279 .. 272
+  uint8_t padding[5];
```
```suggestion
  uint64_t magic = char magic[8] = {'i', 'n', 't', '-', 'd', 'a', 't', 'a'}; // 63 .. 0, must be "int-data"
  /// rvfi_rd_wdata is the value of the x register addressed by rd after
  /// execution of this instruction. This output must be zero when rd is zero.
  uint64_t rvfi_rd_wdata = 0; // 127 ..  64
  /// rvfi_rs1_rdata/rvfi_rs2_rdata is the value of the x register addressed
  /// by rs1/rs2 before execution of this instruction. This output must be
  /// zero when rs1/rs2 is zero.
  uint64_t rvfi_rs1_rdata = 0; // 191 .. 128
  uint64_t rvfi_rs2_rdata = 0; // 255 .. 192
  /// rvfi_rd_addr is the decoded rd register address for the retired
  /// instruction. For an instruction that writes no rd register, this output
  /// must always be zero.
  uint8_t rvfi_rd_addr = 0; // 263 .. 256
  /// rvfi_rs1_addr and rvfi_rs2_addr are the decoded rs1 and rs1 register
  /// addresses for the retired instruction. For an instruction that reads no
  /// rs1/rs2 register, this output can have an arbitrary value. However, if
  /// this output is nonzero then rvfi_rs1_rdata must carry the value stored
  /// in that register in the pre-state.
  uint8_t rvfi_rs1_addr = 0; // 271 .. 264
  uint8_t rvfi_rs2_addr = 0; // 279 .. 272
  uint8_t padding[5] = {0};
```

arichardson commented on c_emulator/rvfi_dii.cpp:
```diff
@@ -16,6 +16,376 @@
+
+void rvfi_zero_exec_packet()
+{
+  rvfi_instruction = Mk_RVFI_DII_Instruction_Packet();
+  rvfi_pc_data = Mk_RVFI_DII_Execution_Packet_PC();
+  rvfi_int_data = Mk_RVFI_DII_Execution_Packet_Ext_Integer();
+  rvfi_inst_data = Mk_RVFI_DII_Execution_Packet_InstMetaData();
+  rvfi_mem_data = Mk_RVFI_DII_Execution_Packet_Ext_MemAccess();
```
```suggestion
  rvfi_instruction = {};
  rvfi_pc_data = {};
  rvfi_int_data = {};
  rvfi_inst_data = {};
  rvfi_mem_data = {};
```

arichardson reviewed (commented):
I like this and it should make it easier to move away from my ugly binary format that was only designed this way since it was a straightfoward extension of the existing one, to something more flexible like CBOR, etc.

Timmmm commented on c_emulator/rvfi_dii_sail.h:
```diff
@@ -0,0 +1,27 @@
+// ****************************************************************************
+// Sail API
+
+mach_bits rvfi_get_cmd(unit);
```
You can use `uint64_t` here - `mach_bits` is an obsolete alias.

Timmmm commented on model/riscv_fetch_rvfi.sail:
```diff
@@ -7,10 +7,10 @@
+  rvfi_set_pc_data_rdata(zero_extend(get_arch_pc()));
+  rvfi_set_inst_data_order(minstret);
+  rvfi_set_inst_data_mode(zero_extend(privLevel_to_bits(cur_privilege)));
+  rvfi_set_inst_data_ixl(zero_extend(misa[MXL]));
```
I was thinking that instead we could add handy accessors to the Sail model and preserve them, then call them from the RVFI code.

Like

```
function get_ixl() -> bits(2) = misa[MXL]
```

```
--c-preserve get_ixl
```

```
rvfi_inst_data.rvfi_ixl = zget_ixl();
```

Apart from that, and Alexander's comments this looks like almost exactly what I was thinking - nice one!

Eventually I think we could replace `rvfi_get_insn()` with a generic instruction forcing method but that can be in a future PR.

arichardson commented on c_emulator/rvfi_dii.cpp:
```diff
@@ -16,6 +16,242 @@
+
+uint64_t rvfi_get_v2_trace_size()
+{
+  uint64_t trace_size = 512;
```
```suggestion
  uint64_t trace_size = sizeof(RVFI_DII_Execution_Packet_V2);
```

arichardson commented on c_emulator/rvfi_dii.cpp:
```diff
@@ -16,6 +16,242 @@
+  if (rvfi_mem_data_present) {
+    trace_size = trace_size + sizeof(RVFI_DII_Execution_Packet_Ext_MemAccess);
+  }
+  return trace_size >> 3; // we have to return bytes not bits
```
```suggestion
  return trace_size >> 3; // we have to return bytes not bits
```
```suggestion
  return trace_size;
```

arichardson commented on c_emulator/rvfi_dii.cpp:
```diff
@@ -124,26 +360,59 @@ void rvfi_handler::get_and_send_packet(packet_reader_fn reader,
   }
 }
 
+RVFI_DII_Instruction_Packet RVFI_DII_Instruction_Packet_from_u64(uint64_t value)
```
```suggestion
RVFI_DII_Instruction_Packet RVFI_DII_Instruction_Packet_from_u64(uint64_t value)
```
```suggestion
RVFI_DII_Instruction_Packet RVFI_DII_Instruction_Packet::from_u64(uint64_t value)
```

Might be nicer as a static function defined in the struct?

arichardson commented on c_emulator/rvfi_dii.h:
```diff
@@ -23,7 +23,222 @@ class rvfi_handler {
+// ****************************************************************************
+
+typedef struct {
+  uint32_t rvfi_insn = 0; // 31 ..  0, Instruction word: 32-bit instruction
```
The formatting of the trailing comments is a bit odd, maybe better to just place them before the field definition instead?

arichardson commented on c_emulator/rvfi_dii.h:
```diff
@@ -23,7 +23,222 @@ class rvfi_handler {
+
+// ****************************************************************************
+
+typedef struct {
```
Since we are using C++ there is no need for typedef struct, we can just use a named struct.
```suggestion
struct RVFI_DII_Execution_Packet_InstMetadata {
```

arichardson reviewed (commented):
Still not managed to get through the entire diff, but here are some further comments.

trdthg commented on c_emulator/rvfi_dii.h:
```diff
@@ -23,7 +23,238 @@ class rvfi_handler {
+// ****************************************************************************
+
+struct RVFI_DII_Execution_Packet_V1 {
+  // [00 - 07] Instruction number: INSTRET value after completion.
```
What about using bytes for all?

arichardson replied:
Sounds good to me.

trdthg commented on c_emulator/rvfi_dii.cpp:
```diff
@@ -16,6 +16,151 @@
+RVFI_DII_Instruction_Packet
+RVFI_DII_Instruction_Packet::from_u64(uint64_t value)
+{
+  return RVFI_DII_Instruction_Packet {
+      .rvfi_insn = (uint32_t)(value & 0xFFFFFFFF),
+      .rvfi_time = (uint16_t)((value >> 32) & 0xFFFF),
+      .rvfi_cmd = (uint8_t)((value >> 48) & 0xFF),
+      .padding = (uint8_t)((value >> 56) & 0xFF),
+  };
```
I tested with `set(CMAKE_CXX_STANDARD 20)` and didn't encounter any issues

trdthg commented on c_emulator/rvfi_dii.cpp:
```diff
@@ -16,6 +16,151 @@
+static RVFI_DII_Execution_Packet_Ext_Integer rvfi_int_data = {};
+static RVFI_DII_Execution_Packet_Ext_MemAccess rvfi_mem_data = {};
+static bool rvfi_mem_data_present = false;
+static bool rvfi_int_data_present = false;
```
You can check this commit [`ccafb45` (#1099)](https://github.com/riscv/sail-riscv/pull/1099/commits/ccafb452bc8479beff51ecbd6b36c5b65bd17d07)

trdthg commented:
Test_RIG has passed, using this branch and the master branch

trdthg commented on model/rvfi_dii.sail:
```diff
@@ -6,260 +6,12 @@
 /*  SPDX-License-Identifier: BSD-2-Clause                                                */
 /*=======================================================================================*/
 
-// "RISC-V Formal Interface - Direct Instruction Injection" support
-// For use with https://github.com/CTSRD-CHERI/TestRIG
```
move this to _rvfi_dii.h_?

Timmmm commented on c_emulator/rvfi_dii.h:
```diff
@@ -1,6 +1,251 @@
+  // rvfi_mem_rdata is the pre-state data read from rvfi_mem_addr.
+  // rvfi_mem_rmask specifies which bytes are valid.
+  // CHERI-extension: widened to 32 bytes to allow reporting 129 bits
+  uint64_t rvfi_mem_rdata[4] = {};
```
Wouldn't `uint8_t rvfi_mem_rdata[32]` make more sense?

Timmmm reviewed (commented):
A few more comments. Getting pretty close I think.

arichardson commented on c_emulator/rvfi_dii.h:
```diff
@@ -1,6 +1,251 @@
+  RVFI_DII_Execution_Packet_PC pc_data = {};
+  // [63 - 55]
+  uint64_t available_fields = 0;
+  // 448, Followed by RVFI_DII_Execution_Packet_Ext_Integer if set
```
These are still in bits

arichardson replied:
Yeah you'd need to say bit xxx since all other numbers are in bytes

arichardson replied:
Or something like byte 64 bit 1

arichardson commented on c_emulator/rvfi_dii.h:
```diff
@@ -20,10 +265,46 @@ class rvfi_handler {
+
+  // Trace States
+  unsigned trace_version = 1;
+
```
I'd make these non-static members so they are only usable when rvfi is enabled. But can also be done as a cleanup later.

trdthg replied:
The problem is that，we have moved all the rvfi_xxx variables to class members, but from the sail side, it can only access regular C functions not class methods (maybe possibly? I don't know, please let me know if possible), so the associated variables/functions must to be static

arichardson replied:
This should be fine we just have to dereference the global instance of the handler

trdthg replied:
Oh, I forget this. But now I have encountered another problem. The rvfi global variable is currently defined in _main.cpp_, which is a bit difficult to handle in the `CMakeLists.txt`

I am considering a solution to completely separate sail_model(lib, don't need rvfi) and c_simulator(exe, need rvfi), main_unit_tests(exe, don't need rvfi).

- set `callbacks` and `sail_rvfi_apis` to empty impl
- introduce a register function, which could be called by c_simulator to replace them to any other implementation.

looks like this:

```cpp
using MemWriteCallback = unit (*)(const char *type, sbits paddr, uint64_t width, lbits value);

// default impl is empty
MemWriteCallback mem_write_callback_impl = nullptr;

// sail api wrapper
unit mem_read_callback(const char *type, sbits paddr, uint64_t width,
                       lbits value)
{
  if (mem_write_callback_impl) {
    mem_write_callback_impl(type, paddr, width, value);
  }
  return UNIT;
}

// register custom impl
void register_callbacks(MemWriteCallback mem_write_cb = nullptr)
{
  mem_write_callback_impl = mem_write_cb;
}
```

then in _c_emulator.cpp_

```
// register real mem_write_callback implementation
register_callbacks(mem_write_callback)
```

arichardson commented on c_emulator/sail_utils.h:
```diff
@@ -0,0 +1,39 @@
+{
+  static_assert(sizeof(T) <= sizeof(uint64_t),
+                "sbits only accepts types that fit in uint64_t.");
+  return {sizeof(T) * 8, static_cast<unsigned int>(val)};
```
```suggestion
  return {sizeof(T) * 8, static_cast<uint64_t>(val)};
```
Shouldn't this be 64 bits rather than truncation?

arichardson commented on c_emulator/sail_utils.h:
```diff
@@ -0,0 +1,39 @@
+void convert_lbits_to_u8s(lbits val, uint8_t *arr, size_t len);
+lbits convert_u8s_to_lbits(const uint8_t *arr, unsigned int num_bits);
+
+template <typename T> inline sbits make_sbits(T val)
```
This is only used for printing with print_sbits(), can't we avoid all this indirection and print the value directly?

arichardson commented on c_emulator/rvfi_dii.cpp:
```diff
@@ -14,8 +14,152 @@
+
+void rvfi_handler::print_rvfi_exec()
+{
+  print_sbits("rvfi_intr     : ", make_sbits(rvfi_inst_data.rvfi_intr));
```
```suggestion
  print_sbits("rvfi_intr     : %#016jx\n", (uintmax_t)rvfi_inst_data.rvfi_intr);
```

How about this instead? Uintmax_t and %j printf modifier to deal with different uint64_t representations. Or you could use the PRIx macros

trdthg replied:
`print_sbits` currently decides whether to print in binary or hex based on the `sbits.len`, and it will only print the length of `sbits.len`. (The code is mainly modified from `print_bits` which is used for mpz_xxx.)

Using `uintmax_t` is obviously not very good?

arichardson replied:
But all the callers here have a bit width that is a multiple of 8, so always print in hex. Printf with the PRI macros seems much easier

trdthg commented on c_emulator/sail_utils.cpp:
```diff
@@ -0,0 +1,52 @@
+
+void convert_lbits_to_u8s(lbits val, uint8_t *arr)
+{
+  mpz_export(arr, NULL, -1, 1, 0, 0, *val.bits);
```
Added a "len"(byte) param to check the `len >= val.len`

arichardson commented on c_emulator/rvfi_dii.cpp:
```diff
@@ -14,8 +14,141 @@
+  // that we support the version 2 wire format. This is required to keep
+  // backwards compatibility with old implementations that do not support
+  // the new trace format.
+  //  res = {};
```
```suggestion
```

arichardson commented on c_emulator/riscv_callbacks_rvfi.cpp:
```diff
@@ -3,41 +3,41 @@
 void rvfi_callbacks::mem_write_callback(const char *, sbits paddr,
                                         uint64_t width, lbits value)
 {
   if (config_enable_rvfi) {
```
Isn't this always true by definition when the callbacks are registered now? I think we can drop the if.

arichardson commented on c_emulator/rvfi_dii.cpp:
```diff
@@ -88,40 +221,25 @@ bool rvfi_handler::setup_socket(bool config_print)
   if (send_size > 4096) {
     fprintf(stderr, "Unexpected large packet size (> 4KB): %zd\n", send_size);
-    exit(EXIT_FAILURE);
+    exit(1);
```
Any reason for this change? Symbolic names are generally cleaner.

arichardson commented on c_emulator/rvfi_dii.cpp:
```diff
@@ -88,40 +221,25 @@ bool rvfi_handler::setup_socket(bool config_print)
-void rvfi_handler::get_and_send_packet(packet_reader_fn reader,
-                                       bool config_print)
+template <typename T>
+void rvfi_handler::send_packet(const T *data, bool config_print)
```
```suggestion
void rvfi_handler::send_packet(const T &data, bool config_print)
```
Since we are writing only one element I would use a reference to show that it's not an array.

arichardson commented on c_emulator/rvfi_dii.cpp:
```diff
@@ -88,40 +221,25 @@ bool rvfi_handler::setup_socket(bool config_print)
+  if (write(dii_sock, data, send_size) != send_size) {
     fprintf(stderr, "Writing RVFI DII trace failed: %s\n", strerror(errno));
-    exit(EXIT_FAILURE);
+    exit(1);
```
Same here, keep EXIT_FAILURE?

arichardson commented on c_emulator/rvfi_dii.h:
```diff
@@ -9,8 +254,6 @@ enum rvfi_prestep_t {
 
-typedef void (*packet_reader_fn)(lbits *rop, unit);
-
 class rvfi_handler {
```
```suggestion
class rvfi_handler : public callbacks_if {
```

I think it would be simpler to implement the callback interface here, then you don't need all the delegation functions in riscv_callbacks_rvfi.

arichardson reviewed (commented):
A few more comments but generally I think we are very close to being ready to merge this.

trdthg commented on c_emulator/rvfi_dii_sail.cpp:
```diff
@@ -0,0 +1,77 @@
+#include "rvfi_dii_sail.h"
+#include <optional>
+
+// Use anonymous namespaces to avoid naming conflicts
+namespace {
+rvfi_handler *rvfi;
+}
+
+void register_rvfi_dii_handler(rvfi_handler *handler)
+{
+  rvfi = handler;
+}
+
```
The reason I did this is that the rvfi variable is defined in c_emulator, and if I use extern, then riscv-model cannot be compiled separately.

I hope there is a way to completely separate rvfi from riscv-model; currently, I only thought of `class rvfi_api_if` (similar to callback_if) can achieve this.

trdthg commented:
Need some help dealing with lean build issue. @ineol 

```
Fatal error: exception Invalid_argument("List.combine")
```

It seems to come from `(List.combine import_files (List.combine import_names_camel import_refs_camel))`

I did my best to do some statistics

- len(import_names_camel) = 101
- len(import_refs_camel) = 102

trdthg commented:
> Is there anything missing from this PR? 

Rebased. I have completed what I intended to do.

Also fixed the Lean build @ineol, it seems like just a minor issue
```
val rvfi_set_inst_data_insn = impure "rvfi_set_inst_data_insn" : bits(64) -> unit
// function rvfi_set_inst_data_insn(_) = () // This line is not needed, I just wrote it wrong
```

arichardson reviewed (commented):
Mostly looking good. IMO we should merge this asap to avoid further conflicts and then improve it post-merge rather than stealing until it is perfect.

trdthg commented on model/postlude/step.sail:
```diff
@@ -245,7 +245,7 @@ function try_step(step_no : nat, exit_wait : bool) -> bool = {
       // TODO: Remove this and implement it in C.
       if get_config_rvfi() then {
-        rvfi_pc_data[rvfi_pc_wdata] = zero_extend(get_arch_pc())
+        rvfi_set_pc_data_wdata(zero_extend(get_arch_pc()))
```
> it should be handled by pc_write_callback?
looks good, also hope to do this in further pr 😇.

arichardson commented on c_emulator/riscv_sim.cpp:
```diff
@@ -613,29 +605,34 @@ int inner_main(int argc, char **argv)
-  uint64_t entry = rvfi ? rvfi->get_entry()
-                        : load_sail(initial_elf_file, /*main_file=*/true);
+  uint64_t entry;
+  if (rvfi) {
```
Probably better to use .has_value()?

Timmmm replied:
Oh yeah I forgot it wasn't a pointer... one of the reasons I don't like implicit bool conversion!
